### DISCUSSION TITLE
New Major Version v5: Set values via HA --> MQTT --> Solmate

### DISCUSSION THREAD

**Discussion by mmattel** [mmattel/EET-Solmate#50] [Announcements]
For details see: [New Major Version v5](https://github.com/mmattel/EET-Solmate/releases/tag/v5.0.0).

You now can trigger changes via HA. Either manually via the sliders, or setting the entitied directly or like via node red writing to the entities 😄

**Comment by lukashunziker99** [comment]
Works great - thank you!!! Is there a button to start boost?

**Reply by mmattel** [reply]
If you look into the Solmates webUI, there is no such "initiator" and the help function does also not tell much about the boost functionality. From my pov, changed values initiate boost. BTW, thanks for your coffee ☕ 😄

**Reply by BetaTobi** [reply]
BTW: Your "buy me a coffee"-link does not work anymore :-)

**Reply by mmattel** [reply]
Uhhhh, thanks for notifying me, have not recognized that - I am on it and will drop a note when fixed.

**Reply by BetaTobi** [reply]
Just saw, that the boost topic is already discussed here.
Anyway, I just added an issue, hope this was fine.

**Reply by mmattel** [reply]
@BetaTobi the BMAC link is working again 😄

**Reply by BetaTobi** [reply]
:-)
Very good --> enjoy your coffee :)

**Comment by josy1024** [comment]
which python version(s) are tested?

**Reply by mmattel** [reply]
I am running my script with `systemd` on a RPi with Ubuntu that has Python 3.9.
With the current implementation in HA, we have a Python virtual env derived from HA which is Python 3.12
Means Python 3.9 - 3.12 are working. Hope this helps.

**Reply by josy1024** [reply]
Running Version 5 about a month (on a seperate Pi Zero) - and is **VERY stable**. :-) 👍 *COOL*

FEEDBACK:

1) FIX: Dashboard config changed names to:
FROM: solmate_battery_flow  TO **solmate_live_battery_flow**

Source: https://github.com/mmattel/EET-Solmate?tab=readme-ov-file#energy-dashboard

      # virtual EET Solmate sensors
      # battery consumption
      # negative values(battery_flow) = charging or 0
      - name: "Solmate faked Battery Consumption"
        unique_id: "solmate_faked_battery_consumption"
        unit_of_measurement: "W"
        device_class: "power"
        icon: "mdi:battery-charging-40"
        state: "{{ -([ 0, states('sensor.solmate_live_battery_flow') | float(0) ] | min) }}"

      # battery production
      # production = positive values(inject_power) or 0
      - name: "Solmate faked Battery Production"
        unique_id: "solmate_faked_battery_production"
        unit_of_measurement: "W"
        device_class: "power"
        icon: "mdi:home-battery-outline"
        state: "{{ ([ 0, states('sensor.solmate_live_battery_flow') | float(0) ] | max) }}"
        

Next steps is changing the set "user" values based on Current GRID "LOAD". (node-red - Grid Load Optimizer)

2) Diagnose is not updateing in HA - do i need "cloud" for this feature? 

    #.env 
    eet_server_uri="ws://192.168.x.19:9124/"
    eet_local_subdomain="192.168.x.19"


![image](https://github.com/mmattel/EET-Solmate/assets/7140712/9b7cf623-0634-46d7-a906-c72c00c3814e)

**Reply by mmattel** [reply]
I will fix pt 1 asap, it is just a text thing. Thanks for the heads up 👍 --> fixed via #57 

> Diagnose is not updateing in HA - do i need "cloud" for this feature?

No, this is independent and works with both. I also have local connection and it shows all items.
Can you do me a favour:
* Show me the log entries in syslog corresponding to solmate
* Can you print me the infos from the solmates webui `#/my/solmate/information` (anonymize sn# and IP)
* What does MQTT explorer tell about the values (eet/sensor/solmate/info). Note that the infos get updated on each restart and every 24h at 23:45.

What you could do after providing the infos is, that you restart the program manually and recheck if the info section gets updated - which it should!

**Reply by josy1024** [reply]
weird interesting... - seams i have is a "local" only solmate issue.

**LOCAL**:
![image](https://github.com/mmattel/EET-Solmate/assets/7140712/5e03a762-c5fd-46ee-bc7c-86126e5118a9)

**CLOUD**:
![image](https://github.com/mmattel/EET-Solmate/assets/7140712/b8f5af84-cacc-4495-b515-9420a985c9ac)

**LOGS**: (/var/log/messages )

> Jun 25 21:14:06 raspigrid solmate.py: Using env file: /home/user/evn/solmate/.env
> Jun 25 21:14:06 raspigrid solmate.py: Initializing websocket connection.
> Jun 25 21:14:06 raspigrid solmate.py: Create Socket.
> Jun 25 21:14:06 raspigrid solmate.py: Websocket is connected to: ws://192.168.x.xx:9124/
> Jun 25 21:14:06 raspigrid solmate.py: Authenticating.
> Jun 25 21:14:08 raspigrid solmate.py: Authentication to S1K0506A0000xxxx successful!
> Jun 25 21:14:08 raspigrid solmate.py: SolMate is online.
> Jun 25 21:14:09 raspigrid solmate.py: Initializing the MQTT class.
> Jun 25 21:14:09 raspigrid solmate.py: Initializing the MQTT client.
> Jun 25 21:14:09 raspigrid solmate.py: MQTT is connected and running.
> Jun 25 21:14:10 raspigrid solmate.py: Update MQTT topics for Homeassistant.
> Jun 25 21:14:10 raspigrid solmate.py: Once a day queries called by scheduler.
> Jun 25 21:14:10 raspigrid solmate.py: Response: {'type': 'BadRequest', 'message': 'Could not get Solmate info.'}
> Jun 25 21:14:10 raspigrid solmate.py: An non breaking error happened, continuing.
> Jun 25 21:14:10 raspigrid solmate.py: Error: Response: {'type': 'BadRequest', 'message': 'Could not get Solmate info.'}

never got an "Info" via MQTT Explorer

**Reply by mmattel** [reply]
Ok, we both have the same SW version on the solmate running which is good. Compared to your webUI output, I get locally connected also the IP address and SW version which is really strange.

EET provides an API you can query the Solmate and a special call is, to query which queries are available and their  responses. I have added the possibility to do exactly that for testing purposes, a kind of hidden feature...

In your `.env` file, at the bottom, there is a variable `general_api_info=False`, if not present add it and set it to `True` (capital T). Then restart the program. Most easiest via the command line for the output on the shell. That one may be a bit longer but please post it here and embedd it into triple backticks (```). The backticks on an own line each. The program ends post printing and afterwards you can set the var to `False` or just comment it and start the program for normal operation. Based on the output I can continue to investigate.

**Reply by josy1024** [reply]
i've been writing to eet an asked if that behavior is normal.


'get_api_info' route info requested:

```
{
  "routes": [
    {
      "route": "live_values",
      "description": "The battery_ﬂow sign is deﬁned to be: positive battery_ﬂow <=> battery is charged by PV | negative battery_ﬂow <=> battery is discharged (-> injection to household/offgrid)",
      "example_response": {
        "timestamp": "2021-09-27T15:57:34",
        "battery_flow": 44.217,
        "inject_power": 28.0,
        "temperature": 24.0,
        "pv_power": 75.8339,
        "battery_state": 0.2
      },
      "example_request": {
        "route": "live_values",
        "data": {},
        "id": 0
      }
    },
    {
      "route": "get_solmate_info",
      "description": "Get information about the SolMate.",
      "example_response": {
        "version": "",
        "ip": "192.168.0.123"
      },
      "example_request": {
        "route": "get_solmate_info",
        "data": {},
        "id": 0
      }
    },
    {
      "route": "get_api_info",
      "description": "Show usage and route description of the API.",
      "example_response": "This response.",
      "example_request": {
        "route": "api_info",
        "data": {},
        "id": 0
      }
    },
    {
      "route": "get_grid_mode",
      "description": "Show if SolMate is in OFFGrid or ONGrid Mode. Returns mode \"on_grid\" or \"off_grid\".",
      "example_response": {
        "mode": "on_grid"
      },
      "example_request": {
        "route": "get_grid_mode",
        "data": {},
        "id": 0
      }
    },
    {
      "route": "get_user_settings",
      "description": "Show your user settings.",
      "example_response": {
        "updated_at": "2021-09-27T14:29:24",
        "user_nickname": "Dein SolMate",
        "postcode": "1230",
        "country_code": "AT"
      },
      "example_request": {
        "route": "get_user_settings",
        "data": {},
        "id": 0
      }
    },
    {
      "route": "get_injection_settings",
      "description": "Show your injection settings.",
      "example_response": {
        "user_maximum_injection": 200,
        "user_minimum_battery_percentage": 0,
        "user_minimum_injection": 30
      },
      "example_request": {
        "route": "get_injection_settings",
        "data": {},
        "id": 0
      }
    },
    {
      "route": "set_user_minimum_injection",
      "description": "Set minimum injection in Watts. Must be less than or equal to 200 Watts.",
      "example_response": {
        "success": true
      },
      "example_request": {
        "route": "set_user_minimum_injection",
        "data": {
          "injection": 10
        },
        "id": 0
      }
    },
    {
      "route": "set_user_maximum_injection",
      "description": "Set maximum injection in Watts. Must be less than or equal to 800 Watts.",
      "example_response": {
        "success": true
      },
      "example_request": {
        "route": "set_user_maximum_injection",
        "data": {
          "injection": 800
        },
        "id": 0
      }
    },
    {
      "route": "set_user_minimum_battery_percentage",
      "description": "Set minimum battery percentage in percent.",
      "example_response": {
        "success": true
      },
      "example_request": {
        "route": "set_user_minimum_battery_percentage",
        "data": {
          "battery_percentage": 12
        },
        "id": 0
      }
    },
    {
      "route": "logs",
      "description": "Route is only available on the Sol server. Get pv power, battery state and injection power within a given timeframe and resolution. A resolution of 2 means that two 15 minute values will be averaged into one. Therefore a resolution of 1 is the most accurate. Start and End time are DateTime objects.",
      "example_response": {
        "logs": [
          {
            "start": "2022-01-01T00:00:00",
            "battery_state": [
              0.5,
              0.4
            ],
            "pv_power": [
              100.0,
              110.0
            ],
            "inject_power": [
              200.0,
              210.0
            ],
            "end": "2022-02-01T23:59:59"
          }
        ]
      },
      "example_request": {
        "route": "logs",
        "data": {
          "timeframes": [
            {
              "start": "2022-01-01T00:00:00",
              "resolution": 4,
              "end": "2022-02-01T23:59:59"
            }
          ]
        },
        "id": 0
      }
    },
    {
      "route": "get_boost_injection",
      "description": "Get settings of the temporary boost injection. The \"actual_wattage\" might differ from the \"set_wattage\" by a few watts because of inverter inaccuracies.",
      "example_response": {
        "data": {
          "set_wattage": 500,
          "actual_wattage": 496,
          "set_time": 7200,
          "remaining_time": 6000
        }
      },
      "example_request": {
        "route": "get_boost_injection",
        "data": {},
        "id": 0
      }
    },
    {
      "route": "set_boost_injection",
      "description": "Set a temporary boost injection to inject \"set_wattage\" for a specific time \"set_time\" in seconds.",
      "example_response": {
        "success": true
      },
      "example_request": {
        "route": "set_boost_injection",
        "data": {
          "set_time": 7200,
          "set_wattage": 500
        },
        "id": 0
      }
    },
    {
      "route": "get_injection_profiles",
      "description": "Get all the injection profiles on the solmate. The \"min\" and \"max\" arrays contain 24 values normalized in range [0, 1].",
      "example_response": {
        "injection_profiles_timestamp": "2024-03-17T21:33:51.852Z",
        "injection_profiles": {
          "$eet_default_h0": {
            "active": false,
            "protected": false,
            "max": [
              0.604,
              0.51,
              0.504,
              0.504,
              0.504,
              0.59,
              0.966,
              1,
              1,
              1,
              1,
              1,
              1,
              1,
              1,
              1,
              1,
              1,
              1,
              1,
              1,
              0.98,
              0.86,
              0.694
            ],
            "name": "$eet_default_h0",
            "min": [
              0.045,
              0.023,
              0.023,
              0.022,
              0.023,
              0.027,
              0.07,
              0.179,
              0.191,
              0.162,
              0.13,
              0.12,
              0.126,
              0.147,
              0.126,
              0.105,
              0.098,
              0.128,
              0.194,
              0.233,
              0.227,
              0.18,
              0.107,
              0.069
            ]
          }
        }
      },
      "example_request": {
        "route": "get_injection_profiles",
        "data": {},
        "id": 0
      }
    },
    {
      "route": "set_injection_profiles",
      "description": "Set the injection profiles on the solmate. Be aware that this will overwrite all the injection profiles on the solmate as long as \"injection_profile_timestamp\" is newer than the local \"injection_profile_timestamp\" on the solmate. The \"min\" and \"max\" arrays each contain 24 values normalized in range [0, 1].",
      "example_response": {
        "success": true
      },
      "example_request": {
        "route": "set_injection_profiles",
        "data": {
          "injection_profiles_timestamp": "2024-03-17T21:33:51.852Z",
          "injection_profiles": {
            "awesome_injection_profile_id": {
              "active": false,
              "protected": false,
              "max": [
                0.604,
                0.51,
                0.504,
                0.504,
                0.504,
                0.59,
                0.966,
                1,
                1,
                1,
                1,
                1,
                1,
                1,
                1,
                1,
                1,
                1,
                1,
                1,
                1,
                0.98,
                0.86,
                0.694
              ],
              "name": "Awesome injection profile",
              "min": [
                0.045,
                0.023,
                0.023,
                0.022,
                0.023,
                0.027,
                0.07,
                0.179,
                0.191,
                0.162,
                0.13,
                0.12,
                0.126,
                0.147,
                0.126,
                0.105,
                0.098,
                0.128,
                0.194,
                0.233,
                0.227,
                0.18,
                0.107,
                0.069
              ]
            }
          }
        },
        "id": 0
      }
    },
    {
      "route": "apply_injection_profile",
      "description": "Apply the injection profile with the id \"id\".",
      "example_response": {
        "success": true
      },
      "example_request": {
        "route": "apply_injection_profile",
        "data": {
          "id": "awesome_injection_profile_id"
        },
        "id": 0
      }
    }
  ],
  "information": "This API is not versioned and routes might change in the future. Please take this in consideration when using it. Every request has to have an ongoing id, to identify the message, a route, as well as a data dictionary."
}

```

**Reply by mmattel** [reply]
Q: have you tested using the local or cloud config?

"postcode": "1230" --> do I read correctly you live in Vienna? We would not be far apart if so...

**Reply by josy1024** [reply]
hm. my postcode is 3622 - lower austria. - i dont know where 1230 comes??

haveing some issues trying to **set "unpossible"** (float) values via mqtt. (i will fix that on my side too)
but it seems that the service shutting complete down. a little part for improvement.


```
Jun 29 00:03:23 raspigrid solmate.py[20154]: WS Request: Connection closed unexpectedly.
Jun 29 00:03:23 raspigrid solmate.py[20154]: Waiting: timer_conn_err: 10s
Jun 29 00:03:33 raspigrid solmate.py[20154]: Shutting down MQTT gracefully.

Jun 29 07:15:11 raspigrid python3[21844]:   File "/home/user/evn/solmate/solmate_mqtt.py", line 329, in _on_message
Jun 29 07:15:11 raspigrid python3[21844]:     write_dict[self.real_names[i]] = int(value)
Jun 29 07:15:11 raspigrid python3[21844]: ValueError: invalid literal for int() with base 10: '128.5'
```

**Reply by mmattel** [reply]
greetings to lower austria, not much distance to 1020 😄 
regarding to set float values, I recently have identified this too and it turns out to be related to different locale settings which is a nightmare. when setting the values via the gui, the issue does not occur because of autoconfig. but when e.g. using node red, you may calulate fractional numbers and they can cause issues when casting to int - dot/comma mixup.

I need to look if that is programatically solvable...
In any case I will, as part of the doc restructuring and improvement, add a note about sending value types.

**Reply by josy1024** [reply]
yes with node red i'm setting min, max, and batt and script is periodicalli hanging ;-(

do you have a stable stress test of the "set values" -  every x seconds set all 3 values (with a 200ms sleep ) - I'm trying to play around not to crash the script.

**Reply by mmattel** [reply]
I am just refactoring the code including setting values from mqtt plus a lot of new stuff. This is not that easy. Regarding setting values, these MUST be integer and MUST NOT contain any dots and commas. There will be an error logged with the key and the value in question. To proceed the process, I will use the former correct value... 
A docs note will be added too. I have not found a general way to overcome this problem.

Note that I highly recommend to have the interval setting values NOT to often. Not because of my code but because the solmate does not react that quickly though the values are immediately set.

**Reply by josy1024** [reply]
hab "int" mit float und round ergänzt, damit das script nicht regelmäßig crashed.
```
solmate_mqtt.py:323:                                            write_dict.update({self.real_names[i]: int(round(float(value)))})
solmate_mqtt.py:330:                                                    write_dict[self.real_names[i]] = int(round(float(value)))
```

**Reply by josy1024** [reply]
some robustness feedback...
as incredible solmate-api works there is some improvement possible on minimal "**environmental**" things:

situation:
this script "hangs" if something "in between" happens...
the script does not recover if ha/mqtt server got a reboot. (installed on seperate pi)

```
     Loaded: loaded (/etc/systemd/system/eet.solmate.service; enabled; vendor p>
     Active: inactive (dead) since Mon 2024-07-01 10:13:28 CEST; 10min ago
Jul 01 10:13:28 raspigrid solmate.py[1093]: Initializing the MQTT class.
Jul 01 10:13:28 raspigrid solmate.py[1093]: Initializing the MQTT client.
Jul 01 10:13:28 raspigrid solmate.py[1093]: MQTT connection failed: [Errno 111] Connection refused
```

how if's solved this: (not only for mqtt, also for the sagem evn grid script)

i'v used some service **watchdog** "**sdnotify**" if something happening (ha-reboot, mqtt reboot)
details: https://github.com/josy1024/eet2mqtt/blob/main/sol2mqtt.py
https://github.com/josy1024/eet2mqtt?tab=readme-ov-file#eetservice

**Comment by mmattel** [comment]
> hab "int" mit float und round ergänzt, damit das script nicht regelmäßig crashed.

das ist in der neuen version gefixed, allerdings in erweitertem masse. es geht darum, dass eine konvertierung auch da fehlschlagen kann und auch das fange ich noch ab und dann wird der letztgültige wert als ersatz genommen. achtung, das steht dann auch im readme, es werden nur zahlen ohne vorzeichen und ohne punkt und komma genommen. du musst das also im node red richtig stellen. das geht leider generell nicht anders, auch wenn es für dich mit deinem fix passen würde...

> MQTT connection failed: [Errno 111] Connection refused

ah, guter punkt. ich fange ja websocket errors ab, aber beim mqtt hab ich wie du sagst noch potential 😅 ich bau da die gleiche logik der fehlertoleranz kein, thx für den hinweis.

**Comment by mmattel** [comment]
@josy1024 mit PR #58 (Updates for version 6) bereite ich die neue version vor. bevor ich das tagge, kannst du gerne master mal auschecken und entsprechend probieren. siehe changelog und breaking changes für infos.

**Reply by josy1024** [reply]
@mmattel zu info: bin jetzt auf 6.1 und die Versions-Infos sind jetzt auch lokal alle da. :-)
bin jetzt komplett auf deinen code geswitched *horray*
auch das werte setzen via nod-red läuft jetzt zuverlässig! 
HA: 2024.7.2 - hatte ein paar Probleme mit der 2024.7.1

**Reply by mmattel** [reply]
@josy1024 freut mich zu hören dass das alles gut klappt 😃
